File: leetcode/977.squares-of-a-sorted-array/cpp/sortedsquares.cpp
#include <vector>
#include <algorithm>

class Solution
{
public:
    std::vector<int> sortedSquares(std::vector<int> &nums)
    {
        return sortedSquaresOptimized(nums);
    }

private:
    // N, N
    std::vector<int> sortedSquaresOptimized(std::vector<int> &nums)
    {
        std::vector<int> arr(nums.size());
        int size = static_cast<int>(nums.size());

        int l = 0;
        int r = size - 1;
        int n = size - 1;

        while (n >= 0)
        {
            int sl = nums[l] * nums[l];
            int sr = nums[r] * nums[r];

            if (sl > sr)
            {
                arr[n] = sl;
                l++;
            }
            else
            {
                arr[n] = sr;
                r--;
            }

            n--;
        }

        return arr;
    }

    // NLogN, N
    std::vector<int> sortedSquaresBruteForce(std::vector<int> &nums)
    {
        std::vector<int> arr(nums.size());
        std::transform(nums.begin(), nums.end(), arr.begin(), &square);
        std::sort(arr.begin(), arr.end());

        return arr;
    }

    static int square(int n)
    {
        return n * n;
    }
};

File: leetcode/977.squares-of-a-sorted-array/ts/sortedsquares.ts
const sortedSquares = (nums: number[]): number[] => {
    return sortedSquaresOptimized(nums)
};

// N, N
const sortedSquaresOptimized = (nums: number[]): number[] => {
    let l = 0
    let r = nums.length - 1
    let n = nums.length - 1

    let arr = new Array<number>(nums.length)

    while (n >= 0) {
        const sl = nums[l] * nums[l]
        const sr = nums[r] * nums[r]

        if (sl > sr) {
            arr[n] = sl
            l++
        } else {
            arr[n] = sr
            r--
        }

        n--
    }

    return arr
}

// NlogN, N
const sortedSquaresBruteForce = (nums: number[]): number[] => {
    return nums.map(n => n * n).sort((a, b) => a - b);
}

export default sortedSquares
